tensorflow_probability/python/experimental/mcmc/sample_fold_test.py

```diff
@@ -0,0 +1,301 @@
+    new_state = tfp.random.rayleigh(self._shape, seed=seed)
+    return new_state, TestTransitionKernelResults(
+        counter_1=previous_kernel_results.counter_1 + 1,
+        counter_2=previous_kernel_results.counter_2 + 2)
+
+  def bootstrap_results(self, current_state):
+    return TestTransitionKernelResults(
+        counter_1=tf.zeros(()), counter_2=tf.zeros(()))
+
+  @property
+  def is_calibrated(self):
+    return self._is_calibrated
+
+
+class TestReducer(tfp.experimental.mcmc.Reducer):
+  """Simple Reducer that (naively) computes the mean"""
+
+  def initialize(self, initial_chain_state=None, initial_kernel_results=None):
+    return tf.zeros((2,))
+
+  def one_step(self, sample, current_state, previous_kernel_results, axis=None):
+    return current_state + tf.convert_to_tensor([1, sample])
+
+  def finalize(self, final_state):
+    return final_state[1] / final_state[0]
+
+
+@test_util.test_all_tf_execution_regimes
+class SampleFoldTest(test_util.TestCase):
+
+  def test_simple_operation(self):
+    fake_kernel = TestTransitionKernel()
+    fake_reducer = TestReducer()
+    reduction_rslt, warm_restart_pkg = tfp.experimental.mcmc.sample_fold(
+        num_steps=5,
+        current_state=0.,
+        kernel=fake_kernel,
+        reducers=fake_reducer,
+    )
+    reduction_rslt, last_sample, innermost_results = self.evaluate([
+        reduction_rslt,
+        warm_restart_pkg[0],
+        warm_restart_pkg[1].inner_results.inner_results
+    ])
+    self.assertEqual(3, reduction_rslt)
+    self.assertEqual(5, last_sample)
+    self.assertEqual(5, innermost_results.counter_1)
+    self.assertEqual(10, innermost_results.counter_2)
+
+  @parameterized.parameters(1., 2.)
+  def test_current_state(self, curr_state):
+    fake_kernel = TestTransitionKernel()
+    fake_reducer = TestReducer()
+    reduction_rslt, warm_restart_pkg = tfp.experimental.mcmc.sample_fold(
+        num_steps=5,
+        current_state=curr_state,
+        kernel=fake_kernel,
+        reducers=fake_reducer,
+    )
+    reduction_rslt, last_sample, innermost_results = self.evaluate([
+        reduction_rslt,
+        warm_restart_pkg[0],
+        warm_restart_pkg[1].inner_results.inner_results
+    ])
+    self.assertEqual(
+        np.mean(np.arange(curr_state + 1, curr_state + 6)), reduction_rslt)
+    self.assertEqual(curr_state + 5, last_sample)
+    self.assertEqual(5, innermost_results.counter_1)
+    self.assertEqual(10, innermost_results.counter_2)
+
+  def test_warm_restart(self):
+    fake_kernel = TestTransitionKernel()
+    fake_reducer = TestReducer()
+    _, warm_restart_pkg = tfp.experimental.mcmc.sample_fold(
+        num_steps=5,
+        current_state=0.,
+        kernel=fake_kernel,
+        reducers=fake_reducer,)
+    # verify `warm_restart_pkg` works as intended
+    reduction_rslt, warm_restart_pkg = tfp.experimental.mcmc.sample_fold(
+        num_steps=5,
+        current_state=warm_restart_pkg[0],
```

Comment:
The fact that you needed to do this [0] suggests to me that a better API would be for `sample_fold` to return three things: the reduction result, the final chain state, and the final kernel results.  That way the pieces of the return value map more cleanly to the arguments that must be supplied for the next call.